### PR TITLE
Template part e2e fix: try and force waiting for button xpath

### PR DESCRIPTION
**Pull request by getdave**
## Description
This e2e test keeps failing

```
FAIL specs/experiments/multi-entity-saving.test.js (46.989 s)
  ● Multi-entity save flow › Post Editor › Save flow should work as expected.
    TimeoutError: waiting for selector `*[data-type="core/template-part"].block-editor-block-list__layout` failed: timeout 30000ms exceeded
      105 | 			);
      106 | 			await createNewButton.click();
    > 107 | 			await page.waitForSelector( activatedTemplatePartSelector );
```

Looking at the failure artifacts it looks like we've clicked the wrong button and activated the "Choose existing" button.

This is because `*[data-type="core/template-part"].block-editor-block-list__layout` will only exist once we've click the `New template part` button.

Save flow should work as expected  2021-07-02T10-56-38

The test selects the "New template part" button by xpath and that resolves to a DOM node with class variant `primary`. However the button variant is set to change to `tertiary` as soon as the "Choose existing" button is rendered. I wonder if this causes the test to click the wrong button?

This PR tries to force waiting for the xpath before clicking. We'll see...

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by Mamaduka** [comment]
Hi, @getdave

How can I help to get this test fixed? I've seen this a few more a lot in the last couple of days.

**Comment by Mamaduka** [comment]
This might be a good candidate for using `puppeteer-testing-library`, see #28380.

**Comment by getdave** [comment]
@Mamaduka So what I _think_ is happening is that when the placeholder for Template Part first loads the `New template part` is a primary button. But then when the Network resolves and the `Choose existing` button appears, the `New template part` becomes a secondary button.

This process means the selector required to click that button changes. I wonder whether because we're selecting the button by XPath Pupeteer has some internal resolution system which caches the selector as `is-primary` and so when the `click` action is sent it is dispatched on that selector. However because the `New template part` is now a secondary button it ends up triggering the `Choose existing` dialog (as shown in the screenshot). @kevin940726 might know about this.

This is largely hypothetical and there could be a simpler explanation. Another option might be to try using a hardcoded testID on the target button and see if that improves the reliability fo the test in the short term with a view to returning and using a more resilient selector in the future.

This is a hard one because I cannot replicate locally and even on CI it's sometimes passing.

File: packages/e2e-tests/specs/experiments/multi-entity-saving.test.js [hunk 1]
```diff
@@ -100,10 +100,13 @@ describe( 'Multi-entity save flow', () => {
 
 			// Add a template part and edit it.
 			await insertBlock( 'Template Part' );
+
+			await page.waitForXPath( createNewButtonSelector );
 			const [ createNewButton ] = await page.$x(
 				createNewButtonSelector
```
Comment (getdave) [inline packages/e2e-tests/specs/experiments/multi-entity-saving.test.js:106]: Let's try updating this selector to use a hardcoded test id or something. If that improves stability then we know my theory is might be correct.

**Comment by Mamaduka** [comment]
@getdave, you're right. The Puppeteer might be holding a different kind of reference internally for that button.

Here's a screencast simulating button switching via Network throttling:

https://user-images.githubusercontent.com/240569/124906676-5ff97700-dff8-11eb-98b3-da81e3add511.mp4

**Comment by getdave** [comment]
Yes that's the only thing I can think...currently occupied with debugging perf tests for 11.0.0 but will look after that is resolved.

Feel free to dig in or take over though 😄

**Comment by kevin940726** [comment]
From the screencast above, I think the reason is because of layout shifting. So the way puppeteer's `click()` works is very much like how the users interact with it.

1. Puppeteer gets the bounding box of the element that is about to get clicked
2. It moves the cursor to the center of that box
3. Fire `mouse.down` then `mouse.up` event

All of the above are asynchronous and independent, so if layout shifting is happening between 2 and 3, it will click on the wrong element.

I think the solution isn't to solve it in test, but rather to fix it in our code. Whatever puppeteer is experiencing is likely to happen on users too.

A solution would be to delay showing the buttons until all requests are resolved, eliminating the intermediate result. Instead, just show a spinner when it's loading, just like how other blocks work. I'm not familiar with template parts blocks though, so no idea if there're any other aspects that I'm not aware of.

**Comment by getdave** [comment]
Aha. Thanks for that knowledge. I had thought the click was triggered programmatically not like a user click.

This is solvable now as you suggest.

**Comment by getdave** [comment]
Fix made elsewhere based on findings here. Closing.

**Pull request closed without merging**